global_controller/global_controller.py

```diff
@@ -0,0 +1,181 @@
+from common.configurable_model import ConfigurableModel
+from bit_vector import BitVector
+from enum import Enum
+import magma as m
+import fault
+
+
+class GC_reg_addr(Enum):
+    TST_addr = 0
+    stall_addr = 1
+    clk_sel_addr = 2
+    rw_delay_sel_addr = 3
+    clk_switch_delay_sel_addr = 4
+
+
+class GC_op(Enum):
+    NOP = 0
+    config_write = 1
+    config_read = 2
+    write_A050 = 4
+    write_TST = 5
+    read_TST = 6
+    global_reset = 7
+    write_stall = 8
+    read_stall = 9
+    advance_clk = 10
+    read_clk_domain = 11
+    switch_clk = 12
+    write_rw_delay_sel = 13
+    read_rw_delay_sel = 14
+    write_clk_switch_delay_sel = 15
+    read_clk_switch_delay_sel = 16
+
+
+def gen_global_controller(config_data_width: int,
+                          config_addr_width: int,
+                          config_op_width: int):
+
+    class _GlobalController(ConfigurableModel(32, 32)):
+        def __init__(self):
+            super().__init__()
+            self.NUM_STALL_DOMAINS = 4
+            self.reset()
+
+        def reset(self):
```

Comment:
would be great to have documentation on the methods, either in line with the file, or in a brief .md file. Just what you think the high level interface to the GC should be. So like

Methods:
- config_read(addr): does some thing
- config_write(addr, data): does some other thing

Attributes:
- reset_out: some output of the GC

I prefer this in a md file rather than inline in this file.